### PR TITLE
per version blocks

### LINKED ISSUE #20748: Allow a non-continuous range of versions to be added to the blocklist

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #20748 

Known deficiencies:
- You can't edit directly from the Block admin page currently, but it was just a shortcut to creating a BlocklistSubmission anyway.  I aim to restore that functionality in a follow-up [#20851]
- admin UI could do with some polish, especially for viewing completed BlocklistSubmission instances
- there aren't enough tests of the new functionality - i.e. being able to create multiple submissions for a single guid at the same time, with differing delays, etc.  I changed the existing tests so I'm pretty confident it works but there should be more so I'll add some more in a follow-up.
- frontend needs updating with the new serializer output, and the dropped fields removed from v5 (behind an API gate) [mozilla/addons-frontend#12252]
- the reviewer page UI when it comes to Blocks and BlockSubmissions is a little inconsistent - but all of that will need changing anyway (again, in a follow-up).

### CODE REVIEW

File: docs/topics/api/blocklist.rst
Comment (eviljeff) [reply docs/topics/api/blocklist.rst:29]: yeah, I just edited the docs - the fields are still present in v5 until I (or someone else) updates frontend and removes them from the API in a follow-up.  Once there are blocks with gaps in the ranges of versions the fields are inaccurate though so I removed them.  I can leave them in the docs for now if you prefer?

**Comment by eviljeff** [comment]
Another set of undefined/sub-optimal behaviour is around block metadata:
- when a guid is submitted in a **multiple** guid blocklistsubmission, but none of that guids versions are selected to be blocked (i.e. all the checkboxes are deselected; or the versions are already present in another blocklistsubmission so not available) currently the reason + url would be updated on that guid's block.
- On the other hand, when a guid is submitted in a **single** guid blocksubmission and none of the versions are selected to be blocked (or all are unavailable), because you _want_ to update the metadata a validation error is raised because at least one version must be selected.
- when there are multiple blocklistsubmissions affecting a guid each submission that is published last will overwrite the previous metadata - e.g. if you _don't_ want to change the reason/url you have to be careful to copy to have the reason/url be the same in each submission, and copy from the existing block

I'm not sure what the solution is to metadata really - maybe a property in the blocklistsubmission instance that indicates that metadata should be updated for that block?  Definitely needs addressing in follow-up.

File: docs/topics/api/blocklist.rst
Comment (diox) [reply docs/topics/api/blocklist.rst:29]: I had not read the code yet so I was a bit surprised. I guess maybe leave them while it's there, with a warning that we have to migrate? We'll probably do it sooner than later but it might be better to be consistent

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (diox) [inline src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:11]: That call will likely fail without `.iterator()` since Django would try to cache all the rows for that queryset. And even with it... That's a lot of rows to process in a migration.

Starting to think we should split this PR into 2:
1. Introduce `BlockVersion`, maintain it as blocks are created/edited/deleted, but do not sure it as the source of truth yet, keep relying on min/max range. Migrate existing blocks using a command that we'd run after the push.
2. Replace min/max range with `BlockVersion` entirely next push (the bulk of what this PR currently contains)

What do you think ?

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (diox) [inline src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:14]: As currently written, if the migration fails mid-way we're in trouble, because it can't pick up where it left off - the `bulk_create()` would fail on already created `BlockVersion`. Could you avoid them in this queryset ?

Comment (eviljeff) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:14]: if we assume that `bulk_create` doesn't fail in a way that some of the `BlockVersion`s are created and not others, then yeah we could filter out any `Block` that already has at least one `BlockVersion`.

Comment (eviljeff) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:14]: Or, as I now think you mean, filter them out in the `File` query.

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (eviljeff) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:11]: > That call will likely fail without .iterator() since Django would try to cache all the rows for that queryset. 

could also make it a `values_list` and only return `id`, `guid`, `min_version`, and `max_version`.

> Starting to think we should split this PR into 2:
>
> 1.     Introduce BlockVersion, maintain it as blocks are created/edited/deleted, but do not sure it as the source of truth yet, keep relying on min/max range. Migrate existing blocks using a command that we'd run after the push.
> 2.     Replace min/max range with BlockVersion entirely next push (the bulk of what this PR currently contains)

It's possible (at some currently unquantifiable amount of effort) - I doubt the second pr will be significantly smaller than this pr is now though.

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (diox) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:14]: Yeah sorry, I meant filter them out of the `File` query.

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (diox) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:11]: > > That call will likely fail without .iterator() since Django would try to cache all the rows for that queryset.
> 
> could also make it a `values_list` and only return `id`, `guid`, `min_version`, and `max_version`.

`.iterator()` also internally does chunking to avoid fetching too much from the database driver layer at the same time, so we likely need it. We could `values_list()` on top of that yeah, but I'm not sure it's that important with `.iterator()`.

**Comment by diox** [comment]
> * On the other hand, when a guid is submitted in a **single** guid blocksubmission and none of the versions are selected to be blocked (or all are unavailable), because you _want_ to update the metadata a validation error is raised because at least one version must be selected.

Yeah I encountered that and it was confusing:
- I had an existing `Block` with 2 versions blocked
- I tried to create a new `BlocklistSubmission` for that same `guid`. It showed me the 2 versions, unchecked. Neither of them are valid choices (and the message about that is confusing - it doesn't tell you why they are not valid choices) so I couldn't submit.

**Comment by eviljeff** [comment]
> > * On the other hand, when a guid is submitted in a **single** guid blocksubmission and none of the versions are selected to be blocked (or all are unavailable), because you _want_ to update the metadata a validation error is raised because at least one version must be selected.
> 
> Yeah I encountered that and it was confusing:
> 
>     * I had an existing `Block` with 2 versions blocked
> 
>     * I tried to create a new `BlocklistSubmission` for that same `guid`. It showed me the 2 versions, unchecked. Neither of them are valid choices (and the message about that is confusing - it doesn't tell you why they are not valid choices) so I couldn't submit.

It's supposed to show you the 2 versions disabled, because they're already blocked, but I guess I introduced a bug along the way.  For a create blocksubmission anyway - a delete blocksubmission is supposed to show existing blocked versions not disabled, so you can unblock them.

File: src/olympia/blocklist/models.py
Comment (diox) [inline src/olympia/blocklist/models.py:289]: Follow-up: I wonder if we still need that logic?

File: src/olympia/blocklist/utils.py
Comment (diox) [inline src/olympia/blocklist/utils.py:190]: Is this logic covered by any tests ? or is it part of the things you were considering in a follow-up ?

**Review by diox — COMMENTED** [review]
Overall this looks good, so well done on the titanic work :) 

I said on slack I was on the fence regarding landing this in 2 stages but now that I'm through reviewing this I think we have to. We will be a lot safer if the first batch we land is completely backwards-compatible, especially if we can do the backfill outside of the migration step of the push.

> there aren't enough tests of the new functionality - i.e. being able to create multiple submissions for a single guid at the same time, with differing delays, etc. I changed the existing tests so I'm pretty confident it works but there should be more so I'll add some more in a follow-up.

I think we'll want those extra tests before merging this, to avoid the potential for edge cases triggered by (accidental or intended) use of the new functionality.

File: src/olympia/blocklist/admin.py
Comment (eviljeff) [reply src/olympia/blocklist/admin.py:344]: Not in this iteration, no.  (Originally I was trying to get the functionality where you edit and save a Block and it creates a  BlockSubmission working again, but I stripped that out because it was too much extra and a nice to have)

File: src/olympia/blocklist/admin.py
Comment (eviljeff) [reply src/olympia/blocklist/admin.py:537]: The follow-up should be restoring the edit functionality on the page, so they'll be checkboxes, but (as I just wrote on the other comment) I had to pull that functionality for now.

File: src/olympia/blocklist/models.py
Comment (eviljeff) [reply src/olympia/blocklist/models.py:289]: We could maybe switch it out for full models, but use `.only()` to load a subset of the data.  Or always load full Block instances and always have a list of `FakeBlockAddonVersion` for addon versions?  Having to handle both cases _does_ add some extra complexity.

File: src/olympia/blocklist/utils.py
Comment (eviljeff) [reply src/olympia/blocklist/utils.py:190]: TestBlockAdminDelete tests this, by deleting one block completely (all blocked versions), and just removing a single blocked version from the other.

File: src/olympia/blocklist/migrations/0031_auto_20230612_1447.py
Comment (eviljeff) [reply src/olympia/blocklist/migrations/0031_auto_20230612_1447.py:11]: resolved in #20837

**Comment by eviljeff** [comment]
merged; addressed tests; added some code and tests! Good news is with the `BlockVersion` changes landing previously there isn't a hairy migration, and some of the already obsolete code isn't cluttering up _this_ PR; Bad news is the extra tests and other changes have _increased_ the changed line cont.

File: src/olympia/blocklist/models.py
Comment (eviljeff) [reply src/olympia/blocklist/models.py:291]: (Note, I didn't address this.  I still could in this PR rather than in a follow-up.)

File: src/olympia/blocklist/admin.py
Comment (eviljeff) [reply src/olympia/blocklist/admin.py:568]: https://github.com/mozilla/addons-server/issues/20851

File: src/olympia/blocklist/models.py
Comment (diox) [reply src/olympia/blocklist/models.py:291]: I was wondering if that was often reached - but I guess it is because `GUID_FULL_LOAD_LIMIT` is not too high.

Comment (diox) [reply src/olympia/blocklist/models.py:291]: Let's keep it like that.

**Review by diox — APPROVED** [review]

**Pull request merged**
